okhttp-testing-support/src/main/kotlin/okhttp3/RecordingConnectionListener.kt

```diff
@@ -0,0 +1,137 @@
+  fun forbidLock(lock: Any) {
+    forbiddenLocks.add(lock)
+  }
+
+  /**
+   * Removes recorded events up to (and including) an event is found whose class equals [eventClass]
+   * and returns it.
+   */
+  fun <T> removeUpToEvent(eventClass: Class<T>): T {
+    val fullEventSequence = eventSequence.toList()
+    try {
+      while (true) {
+        val event = takeEvent()
+        if (eventClass.isInstance(event)) {
+          return eventClass.cast(event)
+        }
+      }
+    } catch (e: NoSuchElementException) {
+      throw AssertionError("full event sequence: $fullEventSequence", e)
+    }
+  }
+
+  /**
+   * Remove and return the next event from the recorded sequence.
+   *
+   * @param eventClass a class to assert that the returned event is an instance of, or null to
+   *     take any event class.
+   * @param elapsedMs the time in milliseconds elapsed since the immediately-preceding event, or
+   *     -1L to take any duration.
+   */
+  fun takeEvent(
+    eventClass: Class<out ConnectionEvent>? = null,
+    elapsedMs: Long = -1L
+  ): ConnectionEvent {
+    val result = eventSequence.remove()
+    val actualElapsedNs = result.timestampNs - (lastTimestampNs ?: result.timestampNs)
+    lastTimestampNs = result.timestampNs
+
+    if (eventClass != null) {
+      assertThat(result).isInstanceOf(eventClass)
+    }
+
+    if (elapsedMs != -1L) {
+      assertThat(
+        TimeUnit.NANOSECONDS.toMillis(actualElapsedNs)
+          .toDouble()
+      )
+        .isCloseTo(elapsedMs.toDouble(), Offset.offset(100.0))
+    }
+
+    return result
+  }
+
+  fun recordedEventTypes() = eventSequence.map { it.name }
+
+  fun clearAllEvents() {
+    while (eventSequence.isNotEmpty()) {
+      takeEvent()
+    }
+  }
+
+  private fun logEvent(e: ConnectionEvent) {
+    if (e.connection != null) {
+      assertThat(Thread.holdsLock(e.connection))
+        .overridingErrorMessage("Called with lock $${e.connection}")
+        .isFalse()
+    }
+    for (lock in forbiddenLocks) {
+      assertThat(Thread.holdsLock(lock))
+        .overridingErrorMessage("Called with lock $lock")
+        .isFalse()
+    }
+
+    eventSequence.offer(e)
+  }
+
+  override fun connectStart(route: Route, call: Call) = logEvent(ConnectionEvent.ConnectStart(System.nanoTime(), route, call))
+
+  override fun connectFailed(route: Route, call: Call, failure: IOException) = logEvent(ConnectionEvent.ConnectFailed(System.nanoTime(), route, call, failure))
+
+  override fun connectEnd(connection: Connection) {
+    logEvent(ConnectionEvent.ConnectEnd(System.nanoTime(), connection))
+  }
+
+  override fun connectionClosed(connection: Connection) = logEvent(ConnectionEvent.ConnectionClosed(System.nanoTime(), connection))
+
+  override fun connectionAcquired(connection: Connection, call: Call) {
+    logEvent(ConnectionEvent.ConnectionAcquired(System.nanoTime(), connection, call))
+  }
+
+  override fun connectionReleased(connection: Connection, call: Call) {
```

Comment:
We should assert that Acquire & Release events are balanced.

Reply:
I can add. Probably as a fast follow.